Controller/ReceivedUrlTrait.php

```diff
@@ -0,0 +1,245 @@
+ */
+
+namespace Bolt\Boltpay\Controller;
+
+use Bolt\Boltpay\Helper\ArrayHelper;
+use Bolt\Boltpay\Helper\Log as LogHelper;
+use Bolt\Boltpay\Helper\Cart as CartHelper;
+use Bolt\Boltpay\Helper\Config as ConfigHelper;
+use Bolt\Boltpay\Helper\Bugsnag;
+use Magento\Framework\Exception\LocalizedException;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Checkout\Model\Session as CheckoutSession;
+use Magento\Quote\Model\Quote;
+use Magento\Sales\Model\Order;
+use Bolt\Boltpay\Helper\Order as OrderHelper;
+
+trait ReceivedUrlTrait
+{
+    /**
+     * @var LogHelper
+     */
+    private $logHelper;
+    /**
+     * @var ConfigHelper
+     */
+    private $configHelper;
+    /**
+     * @var Bugsnag
+     */
+    private $bugsnag;
+    /**
+     * @var CartHelper
+     */
+    private $cartHelper;
+    /**
+     * @var CheckoutSession
+     */
+    private $checkoutSession;
+    /**
+     * @var OrderHelper
+     */
+    private $orderHelper;
+
+    /**
+     * @return \Magento\Framework\App\ResponseInterface
+     */
+    public function execute()
+    {
+        $boltSignature = $this->getRequest()->getParam('bolt_signature');
+        $boltPayload = $this->getRequest()->getParam('bolt_payload');
+        $storeId = $this->getRequest()->getParam('store_id');
+
+        $signature = base64_decode($boltSignature);
+
+        $signingSecret = $this->configHelper->getSigningSecret($storeId);
+
+        $hashBoltPayloadWithKey = hash_hmac('sha256', $boltPayload, $signingSecret, true);
+        $hash = base64_encode($hashBoltPayloadWithKey);
+
+        if ($signature === $hash) {
+            try {
+                $payload = base64_decode($boltPayload);
+                $payloadArray = json_decode($payload, true);
+                $incrementId = $this->getIncrementIdFromPayload($payloadArray);
+
+                /** @var Order $order */
+                $order = $this->getOrderByIncrementId($incrementId);
+
+                /** @var Quote $quote */
+                $quote = $this->getQuoteById($order->getQuoteId());
+
+                $redirectUrl = $this->getRedirectUrl($order);
+
+                // add quote information to the session
+                $this->clearQuoteSession($quote);
+
+                // add order information to the session
+                $this->clearOrderSession($order, $redirectUrl);
+
+                if ($order->getState() == Order::STATE_NEW) {
+                    // Save reference to the Bolt transaction with the order
+                    $order->addStatusHistoryComment(
+                        __(
+                            'Bolt transaction: %1',
+                            $this->orderHelper->formatReferenceUrl($this->getReferenceFromPayload($payloadArray))
+                        )
+                    );
+                    $order->save();
+                } else {
+                    $this->bugsnag->notifyError(
```

Comment:
it's possible order is approved by the time user redirects? not sure if we should treat it as error

Reply:
@daisy1754 i can reproduce it in slow internet connection or if we have some slow staging server.
Sometimes it can take about ~10 seconds and then we can caught such error.
I think we can leave it here to research our and merchant's server bottle necks in productivity way.

Reply:
@daisy1754 we just log the event and continue execution